spec/p2p/v0.34/README.md

```diff
@@ -0,0 +1,140 @@
+
+The PEX reactor ensures a node is connected to peers by running a routine to keep dialing peers in the background until a threshold of connected peers is reached (`MaxNumOutboundPeers`). This way, if a peers' address has been added via an incoming request, it will eventually be dialed. Peers to dial are chosen with some configurable bias for unvetted peers. The bias should be lower when we have fewer peers and can increase as we obtain more, ensuring that our first peers are more trustworthy, but always giving us the chance to discover new good peers. 
+
+As the number of outgoing peers is limited, the reactor will choose the number of addresses to dial taking into account the current number of outgoing connections and ongoing dialings. The addresses to dial will be picked with a bias towards new and vetted peers (TODO define ).
+
+Except for persistent peers, all other peers can be dialed at most `maxAttemptsToDial`. If a node is not connected by that time it is marked as bad. Otherwise, if the dial fails, the node will wait before the next dial. (exponential backoff mechanism). 
+
+If a node needs more peers, the PEX reactor checks first whether peers marked as bad can be reinstated and then also picks a random peer from the store to dial. The PEX reactor relies on the switch to do the actual dialing of a peer.
+
+Once the addresses to dial are known, they are forwarded to the `DialPeersAsync` routine of the switch. Each address is then dialed in parallel and the corresponding peer is added to the `dialing` list. 
+
+**Note** When dialing a peer, the peers' address is not added into the address book as it had to be there in order for the node to dial it in the first place. 
+
+
+#### *Successful dialing*
+
+When a peer is successfully dialed, it is removed from the `dialing` list and added to the `peers` list. The switch then calls the `InitPeer` and `AddPeer` routines of all the reactors that have registered to it. 
+
+Once a peer is successfully dialed, we ask it to provide us with more peers.
+
+#### *Dialing failed*
+
+
+### *Removing peers*
+
+Each reactor can call the `StopPeerForError` method of the switch with the ID of the peer that needs to be removed. Then the switch handles stopping the peer (closing the connection to it), and calls the `RemovePeer` method of all reactors registered to the switch.
+
+## Peer ranking
+
+In v0.34 there is no explicit ranking of peers. When choosing peers to dial, there is slight bias towards new and vetted peers. The amount of bias is higher when there are more peers connected to a node. 
+
+In addition to this, peers  can be marked as bad and removed entirely from the potential candidate list. Reactors themselves can also mark peers as bad or good and thus influence the behaviour of the p2p layer. In v0.34 a peer is marked good only from the consensus reactor whenever a peer delivers a correct consensus message (TODO check conditiosn for this). 
```

Comment:
I think the peer have to forward X consensus messages that are considered by the node. By considered here, I mean that are valid and not duplicated (yet received).